File: code/stable/glassbr/src/csharp/ReadTable.cs
/** \file ReadTable.cs
    \author Nikitha Krithnan and W. Spencer Smith
    \brief Provides a function for reading glass ASTM data
*/
using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;

public class ReadTable {
    
    /** \brief Reads glass ASTM data from a file with the given file name
        \param filename name of file with x y and z data
        \param z_vector list of z values
        \param x_matrix lists of x values at different z values
        \param y_matrix lists of y values at different z values
    */
    public static void func_read_table(string filename, List<double> z_vector, List<List<double>> x_matrix, List<List<double>> y_matrix) {
        StreamWriter outfile;
        outfile = new StreamWriter("log.txt", true);
        outfile.WriteLine("function func_read_table called with inputs: {");
        outfile.Write("  filename = ");
        outfile.Write(filename);
        outfile.WriteLine(", ");
        outfile.Write("  z_vector = ");
        outfile.Write("[");
        for (int list_i1 = 0; list_i1 < z_vector.Count - 1; list_i1++) {
            outfile.Write(z_vector[list_i1]);
            outfile.Write(", /f ");
        }
        if (z_vector.Count > 0) {
            outfile.Write(z_vector[z_vector.Count - 1]);
        }
        outfile.Write("]");
        outfile.WriteLine(", ");
        outfile.Write("  x_matrix = ");
        outfile.Write("[");
        for (int list_i2 = 0; list_i2 < x_matrix.Count - 1; list_i2++) {
            outfile.Write("[");
            for (int list_i1 = 0; list_i1 < x_matrix[list_i2].Count - 1; list_i1++) {
                outfile.Write(x_matrix[list_i2][list_i1]);
                outfile.Write(", /f ");
            }
            if (x_matrix[list_i2].Count > 0) {
                outfile.Write(x_matrix[list_i2][x_matrix[list_i2].Count - 1]);
            }
            outfile.Write("]");
            outfile.Write(", /f ");
        }
        if (x_matrix.Count > 0) {
            outfile.Write("[");
            for (int list_i1 = 0; list_i1 < x_matrix[x_matrix.Count - 1].Count - 1; list_i1++) {
                outfile.Write(x_matrix[x_matrix.Count - 1][list_i1]);
                outfile.Write(", /f ");
            }
            if (x_matrix[x_matrix.Count - 1].Count > 0) {
                outfile.Write(x_matrix[x_matrix.Count - 1][x_matrix[x_matrix.Count - 1].Count - 1]);
            }
            outfile.Write("]");
        }
        outfile.Write("]");
        outfile.WriteLine(", ");
        outfile.Write("  y_matrix = ");
        outfile.Write("[");
        for (int list_i2 = 0; list_i2 < y_matrix.Count - 1; list_i2++) {
            outfile.Write("[");
            for (int list_i1 = 0; list_i1 < y_matrix[list_i2].Count - 1; list_i1++) {
                outfile.Write(y_matrix[list_i2][list_i1]);
                outfile.Write(", /f ");
            }
            if (y_matrix[list_i2].Count > 0) {
                outfile.Write(y_matrix[list_i2][y_matrix[list_i2].Count - 1]);
            }
            outfile.Write("]");
            outfile.Write(", /f ");
        }
        if (y_matrix.Count > 0) {
            outfile.Write("[");
            for (int list_i1 = 0; list_i1 < y_matrix[y_matrix.Count - 1].Count - 1; list_i1++) {
                outfile.Write(y_matrix[y_matrix.Count - 1][list_i1]);
                outfile.Write(", /f ");
            }
            if (y_matrix[y_matrix.Count - 1].Count > 0) {
                outfile.Write(y_matrix[y_matrix.Count - 1][y_matrix[y_matrix.Count - 1].Count - 1]);
            }
            outfile.Write("]");
        }
        outfile.WriteLine("]");
        outfile.WriteLine("  }");
        outfile.Close();
        
        StreamReader infile;
        string line;
        List<string> linetokens = new List<string>(0);
        List<string> lines = new List<string>(0);
        infile = new StreamReader(filename);
        line = (infile.ReadLine());
        linetokens = new List<string>(line.Split(','));
        for (int stringlist_i = 0; stringlist_i < linetokens.Count / 1; stringlist_i += 1) {
            z_vector.Add(Double.Parse(linetokens[stringlist_i * 1 + 0]));
        }
        outfile = new StreamWriter("log.txt", true);
        outfile.Write("var 'z_vector' assigned to ");
        outfile.Write("[");
        for (int list_i1 = 0; list_i1 < z_vector.Count - 1; list_i1++) {
            outfile.Write(z_vector[list_i1]);
            outfile.Write(", /f ");
        }
        if (z_vector.Count > 0) {
            outfile.Write(z_vector[z_vector.Count - 1]);
        }
        outfile.Write("]");
        outfile.WriteLine(" in module ReadTable");
        outfile.Close();
        while (!(infile.EndOfStream)) {
            lines.Add(infile.ReadLine());
        }
        for (int i = 0; i < lines.Count; i += 1) {
            linetokens = new List<string>(lines[i].Split(','));
            List<double> x_matrix_temp = new List<double> {};
            List<double> y_matrix_temp = new List<double> {};
            for (int stringlist_i = 0; stringlist_i < linetokens.Count / 2; stringlist_i += 1) {
                x_matrix_temp.Add(Double.Parse(linetokens[stringlist_i * 2 + 0]));
                y_matrix_temp.Add(Double.Parse(linetokens[stringlist_i * 2 + 1]));
            }
            x_matrix.Add(x_matrix_temp);
            y_matrix.Add(y_matrix_temp);
        }
        outfile = new StreamWriter("log.txt", true);
        outfile.Write("var 'x_matrix' assigned to ");
        outfile.Write("[");
        for (int list_i2 = 0; list_i2 < x_matrix.Count - 1; list_i2++) {
            outfile.Write("[");
            for (int list_i1 = 0; list_i1 < x_matrix[list_i2].Count - 1; list_i1++) {
                outfile.Write(x_matrix[list_i2][list_i1]);
                outfile.Write(", /f ");
            }
            if (x_matrix[list_i2].Count > 0) {
                outfile.Write(x_matrix[list_i2][x_matrix[list_i2].Count - 1]);
            }
            outfile.Write("]");
            outfile.Write(", /f ");
        }
        if (x_matrix.Count > 0) {
            outfile.Write("[");
            for (int list_i1 = 0; list_i1 < x_matrix[x_matrix.Count - 1].Count - 1; list_i1++) {
                outfile.Write(x_matrix[x_matrix.Count - 1][list_i1]);
                outfile.Write(", /f ");
            }
            if (x_matrix[x_matrix.Count - 1].Count > 0) {
                outfile.Write(x_matrix[x_matrix.Count - 1][x_matrix[x_matrix.Count - 1].Count - 1]);
            }
            outfile.Write("]");
        }
        outfile.Write("]");
        outfile.WriteLine(" in module ReadTable");
        outfile.Close();
        outfile = new StreamWriter("log.txt", true);
        outfile.Write("var 'y_matrix' assigned to ");
        outfile.Write("[");
        for (int list_i2 = 0; list_i2 < y_matrix.Count - 1; list_i2++) {
            outfile.Write("[");
            for (int list_i1 = 0; list_i1 < y_matrix[list_i2].Count - 1; list_i1++) {
                outfile.Write(y_matrix[list_i2][list_i1]);
                outfile.Write(", /f ");
            }
            if (y_matrix[list_i2].Count > 0) {
                outfile.Write(y_matrix[list_i2][y_matrix[list_i2].Count - 1]);
            }
            outfile.Write("]");
            outfile.Write(", /f ");
        }
        if (y_matrix.Count > 0) {
            outfile.Write("[");
            for (int list_i1 = 0; list_i1 < y_matrix[y_matrix.Count - 1].Count - 1; list_i1++) {
                outfile.Write(y_matrix[y_matrix.Count - 1][list_i1]);
                outfile.Write(", /f ");
            }
            if (y_matrix[y_matrix.Count - 1].Count > 0) {
                outfile.Write(y_matrix[y_matrix.Count - 1][y_matrix[y_matrix.Count - 1].Count - 1]);
            }
            outfile.Write("]");
        }
        outfile.Write("]");
        outfile.WriteLine(" in module ReadTable");
        outfile.Close();
        infile.Close();
    }
}

File: code/stable/glassbr/src/cpp/ReadTable.cpp
#include "ReadTable.hpp"

#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using std::ifstream;
using std::ofstream;
using std::string;
using std::vector;

void read_table(string filename, vector<double> &z_vect3DSor, vector<vector<double>> &x_matrix, vector<vector<double>> &y_matrix) {
    ofstream outfile;
    outfile.open("log.txt", std::fstream::app);
    outfile << "function read_table called with inputs: {" << std::endl;
    outfile << "  filename = ";
    outfile << filename;
    outfile << ", " << std::endl;
    outfile << "  z_vect3DSor = ";
    outfile << "[";
    for (int list_i1 = 0; list_i1 < (int)(z_vect3DSor.size()) - 1; list_i1++) {
        outfile << z_vect3DSor.at(list_i1);
        outfile << ", ";
    }
    if ((int)(z_vect3DSor.size()) > 0) {
        outfile << z_vect3DSor.at((int)(z_vect3DSor.size()) - 1);
    }
    outfile << "]";
    outfile << ", " << std::endl;
    outfile << "  x_matrix = ";
    outfile << "[";
    for (int list_i2 = 0; list_i2 < (int)(x_matrix.size()) - 1; list_i2++) {
        outfile << "[";
        for (int list_i1 = 0; list_i1 < (int)(x_matrix.at(list_i2).size()) - 1; list_i1++) {
            outfile << x_matrix.at(list_i2).at(list_i1);
            outfile << ", ";
        }
        if ((int)(x_matrix.at(list_i2).size()) > 0) {
            outfile << x_matrix.at(list_i2).at((int)(x_matrix.at(list_i2).size()) - 1);
        }
        outfile << "]";
        outfile << ", ";
    }
    if ((int)(x_matrix.size()) > 0) {
        outfile << "[";
        for (int list_i1 = 0; list_i1 < (int)(x_matrix.at((int)(x_matrix.size()) - 1).size()) - 1; list_i1++) {
            outfile << x_matrix.at((int)(x_matrix.size()) - 1).at(list_i1);
            outfile << ", ";
        }
        if ((int)(x_matrix.at((int)(x_matrix.size()) - 1).size()) > 0) {
            outfile << x_matrix.at((int)(x_matrix.size()) - 1).at((int)(x_matrix.at((int)(x_matrix.size()) - 1).size()) - 1);
        }
        outfile << "]";
    }
    outfile << "]";
    outfile << ", " << std::endl;
    outfile << "  y_matrix = ";
    outfile << "[";
    for (int list_i2 = 0; list_i2 < (int)(y_matrix.size()) - 1; list_i2++) {
        outfile << "[";
        for (int list_i1 = 0; list_i1 < (int)(y_matrix.at(list_i2).size()) - 1; list_i1++) {
            outfile << y_matrix.at(list_i2).at(list_i1);
            outfile << ", ";
        }
        if ((int)(y_matrix.at(list_i2).size()) > 0) {
            outfile << y_matrix.at(list_i2).at((int)(y_matrix.at(list_i2).size()) - 1);
        }
        outfile << "]";
        outfile << ", ";
    }
    if ((int)(y_matrix.size()) > 0) {
        outfile << "[";
        for (int list_i1 = 0; list_i1 < (int)(y_matrix.at((int)(y_matrix.size()) - 1).size()) - 1; list_i1++) {
            outfile << y_matrix.at((int)(y_matrix.size()) - 1).at(list_i1);
            outfile << ", ";
        }
        if ((int)(y_matrix.at((int)(y_matrix.size()) - 1).size()) > 0) {
            outfile << y_matrix.at((int)(y_matrix.size()) - 1).at((int)(y_matrix.at((int)(y_matrix.size()) - 1).size()) - 1);
        }
        outfile << "]";
    }
    outfile << "]" << std::endl;
    outfile << "  }" << std::endl;
    outfile.close();
    
    ifstream infile;
    string line;
    vector<string> linetokens(0);
    vector<string> lines(0);
    infile.open(filename, std::fstream::in);
    std::getline(infile, line);
    linetokens.clear();
    std::stringstream ss;
    ss.str(line);
    string word;
    while (std::getline(ss, word, ',')) {
        linetokens.push_back(word);
    }
    for (int i = 0; i < (int)(linetokens.size()) / 1; i += 1) {
        z_vect3DSor.push_back(std::stod(linetokens.at(i * 1 + 0)));
    }
    outfile.open("log.txt", std::fstream::app);
    outfile << "var 'z_vect3DSor' assigned ";
    outfile << "[";
    for (int list_i1 = 0; list_i1 < (int)(z_vect3DSor.size()) - 1; list_i1++) {
        outfile << z_vect3DSor.at(list_i1);
        outfile << ", ";
    }
    if ((int)(z_vect3DSor.size()) > 0) {
        outfile << z_vect3DSor.at((int)(z_vect3DSor.size()) - 1);
    }
    outfile << "]";
    outfile << " in module ReadTable" << std::endl;
    outfile.close();
    string nextLine;
    while (std::getline(infile, nextLine)) {
        lines.push_back(nextLine);
    }
    for (int i = 0; i < (int)(lines.size()); i += 1) {
        linetokens.clear();
        std::stringstream ss;
        ss.str(lines.at(i));
        string word;
        while (std::getline(ss, word, ',')) {
            linetokens.push_back(word);
        }
        vector<double> x_matrix_temp{};
        vector<double> y_matrix_temp{};
        for (int j = 0; j < (int)(linetokens.size()) / 2; j += 1) {
            x_matrix_temp.push_back(std::stod(linetokens.at(j * 2 + 0)));
            y_matrix_temp.push_back(std::stod(linetokens.at(j * 2 + 1)));
        }
        x_matrix.push_back(x_matrix_temp);
        y_matrix.push_back(y_matrix_temp);
    }
    outfile.open("log.txt", std::fstream::app);
    outfile << "var 'x_matrix' assigned ";
    outfile << "[";
    for (int list_i2 = 0; list_i2 < (int)(x_matrix.size()) - 1; list_i2++) {
        outfile << "[";
        for (int list_i1 = 0; list_i1 < (int)(x_matrix.at(list_i2).size()) - 1; list_i1++) {
            outfile << x_matrix.at(list_i2).at(list_i1);
            outfile << ", ";
        }
        if ((int)(x_matrix.at(list_i2).size()) > 0) {
            outfile << x_matrix.at(list_i2).at((int)(x_matrix.at(list_i2).size()) - 1);
        }
        outfile << "]";
        outfile << ", ";
    }
    if ((int)(x_matrix.size()) > 0) {
        outfile << "[";
        for (int list_i1 = 0; list_i1 < (int)(x_matrix.at((int)(x_matrix.size()) - 1).size()) - 1; list_i1++) {
            outfile << x_matrix.at((int)(x_matrix.size()) - 1).at(list_i1);
            outfile << ", ";
        }
        if ((int)(x_matrix.at((int)(x_matrix.size()) - 1).size()) > 0) {
            outfile << x_matrix.at((int)(x_matrix.size()) - 1).at((int)(x_matrix.at((int)(x_matrix.size()) - 1).size()) - 1);
        }
        outfile << "]";
    }
    outfile << "]";
    outfile << " in module ReadTable" << std::endl;
    outfile.close();
    outfile.open("log.txt", std::fstream::app);
    outfile << "var 'y_matrix' assigned ";
    outfile << "[";
    for (int list_i2 = 0; list_i2 < (int)(y_matrix.size()) - 1; list_i2++) {
        outfile << "[";
        for (int list_i1 = 0; list_i1 < (int)(y_matrix.at(list_i2).size()) - 1; list_i1++) {
            outfile << y_matrix.at(list_i2).at(list_i1);
            outfile << ", ";
        }
        if ((int)(y_matrix.at(list_i2).size()) > 0) {
            outfile << y_matrix.at(list_i2).at((int)(y_matrix.at(list_i2).size()) - 1);
        }
        outfile << "]";
        outfile << ", ";
    }
    if ((int)(y_matrix.size()) > 0) {
        outfile << "[";
        for (int list_i1 = 0; list_i1 < (int)(y_matrix.at((int)(y_matrix.size()) - 1).size()) - 1; list_i1++) {
            outfile << y_matrix.at((int)(y_matrix.size()) - 1).at(list_i1);
            outfile << ", ";
        }
        if ((int)(y_matrix.at((int)(y_matrix.size()) - 1).size()) > 0) {
            outfile << y_matrix.at((int)(y_matrix.size()) - 1).at((int)(y_matrix.at((int)(y_matrix.size()) - 1).size()) - 1);
        }
        outfile << "]";
    }
    outfile << "]";
    outfile << " in module ReadTable" << std::endl;
    outfile.close();
    infile.close();
}
